### PR TITLE
Multi-filesystem followups

### PR DESCRIPTION
**Pull request by jcsp**
This is a tidyup for https://github.com/ceph/ceph/pull/6953

Failing to pass features through to FSMap::encode has no effect a the moment, because it doesn't have any feature flags, but to avoid this being a gotcha in future we should pass features through.

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
@jcsp, can you rebase this? I think it's just the last commit that's for real.

**Comment by gregsfortytwo** [comment]
@jcsp this broke cmake build. http://gitbuilder.sepia.ceph.com/gitbuilder-ceph-tarball-trusty-amd64-cmake/log.cgi?log=c21e3a49fb5fdcfe2cb63d655c4eda39a9c1c9bd

>In file included from /srv/autobuild-ceph/gitbuilder.git/build/src/include/types.h:18:0,
from /srv/autobuild-ceph/gitbuilder.git/build/src/include/utime.h:23,
from /srv/autobuild-ceph/gitbuilder.git/build/src/include/cephfs/libcephfs.h:19,
from cephfs.cpp:361:
/srv/autobuild-ceph/gitbuilder.git/build/src/include/int_types.h:4:22: fatal error: acconfig.h: No such file or directory
#include "acconfig.h"

**Comment by jcsp** [comment]
The cmake breakage will be because the branch had my hack for cython in it (I remove it when I'm done hacking)

**Comment by gregsfortytwo** [comment]
Testing passed: http://pulpito.ceph.com/gregf-2016-03-15_04:07:43-fs-greg-fs-testing-314---basic-mira/
Still pending review.

**Comment by jcsp** [comment]
This PR targeted for master, but we'll want to get at least some of this stuff into Jewel (the "mds: fix mds_info_t::dump" commit especially is a clear bug).

**Comment by jcsp** [comment]
@gregsfortytwo ready for review now

File: src/messages/MFSMap.h [hunk 1]
```diff
@@ -26,16 +26,18 @@ class MFSMap : public Message {
...
 
   MFSMap() : 
     Message(CEPH_MSG_FS_MAP), epoch(0) {}
-  MFSMap(const uuid_d &f, FSMap *fsmap) :
-    Message(CEPH_MSG_FS_MAP), epoch(fsmap->get_epoch())
+  MFSMap(const uuid_d &f, const FSMap &fsmap_) :
+    Message(CEPH_MSG_FS_MAP), epoch(fsmap_.get_epoch())
   {
-    fsmap->encode(encoded, -1);
+    fsmap = fsmap_;
```
Comment (gregsfortytwo) [inline src/messages/MFSMap.h]: Hmm. Do we have any reason to think the fsmap reference will still be valid by the time we get around to encoding it?

Comment (gregsfortytwo) [inline src/messages/MFSMap.h]: ...wow did I read that wrong. Never mind.

File: src/messages/MFSMap.h [hunk 2]
```diff
@@ -48,11 +50,11 @@ class MFSMap : public Message {
   void decode_payload() {
     bufferlist::iterator p = payload.begin();
     ::decode(epoch, p);
-    ::decode(encoded, p);
+    ::decode(fsmap, p);
   }
   void encode_payload(uint64_t features) {
     ::encode(epoch, payload);
-    ::encode(encoded, payload);
+    ::encode(fsmap, payload, features);
```
Comment (gregsfortytwo) [inline src/messages/MFSMap.h]: So if we're changing the wire encoding like this it certainly needs to go in Jewel. Don't think that will be a problem as we're still merging master in to Jewel; sound good to you?

File: src/mon/MDSMonitor.cc [hunk 3]
```diff
@@ -144,7 +144,9 @@ void MDSMonitor::update_from_paxos(bool *need_bootstrap)
   // new map
   dout(4) << "new map" << dendl;
   print_map(fsmap, 0);
-  fsmap.sanity();
+  if (!g_conf->mon_mds_skip_sanity) {
+    fsmap.sanity();
+  }
```
Comment (gregsfortytwo) [inline src/mon/MDSMonitor.cc]: We want to do the config check in FSMap itself; Don't Repeat Yourself and that way any future callers don't need to remember to check it.

**Comment by gregsfortytwo** [comment]
So the other follow-up I was interested in and should be done quickly is that we reproduce the entire MDSMap::decode() function in FSMap::decode(). It seems to be working now but is awfully fragile in that state. Can you whip up a patch that makes it reference MDSMap::decode instead?
(So actually because of the ENCODE_START macros you'll want to put the real work into an MDSMap::_decode() function which takes struct_v as well as the iterator&, and then call into that from both decodes, but you get the idea.)

File: src/mon/MDSMonitor.cc (hunk 3 above)
Comment (jcsp) [inline src/mon/MDSMonitor.cc]: I decided that repeating it in MDSMonitor was the lesser of two evils compared with passing a reference to the config object through into FSMap

File: src/messages/MFSMap.h (hunk 2 above)
Comment (jcsp) [inline src/messages/MFSMap.h]: I didn't think passing explicit features arguments (vs. how it was before) to ::encode actually changed the encoding on the wire, does it?

Obviously changing the actual feature bits changes things, but this is just changing it to pass through the globally configured features as opposed to having a hardcoded set further down.

**Comment by jcsp** [comment]
I haven't forgotten about refactoring the decode method, but let's save that for another PR

File: src/messages/MFSMap.h (hunk 2 above)
Comment (gregsfortytwo) [inline src/messages/MFSMap.h]: I wasn't talking about using the feature bits. There's a difference between encoding an FSMap into the payload, and encoding a bufferlist (of an encoded FSMap) into the payload. (There is the usual wrapping around a bufferlist, right? I don't think it's just an append.)

Comment (jcsp) [inline src/messages/MFSMap.h]: Oh yes, quite right.  We're dropping the leading four bytes of the bl's length.

File: src/mon/MDSMonitor.cc (hunk 3 above)
Comment (gregsfortytwo) [inline src/mon/MDSMonitor.cc]: Mmm, okay. Sometimes I miss having globals you can reference gratuitously. ;)

**Comment by gregsfortytwo** [comment]
http://pulpito.ceph.com/gregf-2016-03-23_23:07:15-fs-greg-fs-testing-323-3---basic-mira/ and http://pulpito.ceph.com/gregf-2016-03-24_15:09:06-fs-greg-fs-testing-323-3---basic-mira/

**Pull request merged**